tools/donate-cpu-server.py

```diff
@@ -142,37 +143,55 @@ def crashReport(results_path):
             continue
         datestr = ''
-        for line in open(filename, 'rt'):
-            line = line.strip()
-            if line.startswith('cppcheck: '):
-                if OLD_VERSION not in line:
-                    # Package results seem to be too old, skip
+        with open(filename, 'rt') as file:
+            for line in file:
+                line = line.strip()
+                if line.startswith('cppcheck: '):
+                    if OLD_VERSION not in line:
+                        # Package results seem to be too old, skip
+                        break
+                    else:
+                        # Current package, parse on
+                        continue
+                if line.startswith(str(current_year) + '-') or line.startswith(str(current_year - 1) + '-'):
+                    datestr = line
+                if line.startswith('count:'):
+                    if line.find('Crash') < 0:
+                        break
+                    package = filename[filename.rfind('/')+1:]
+                    counts = line.strip().split(' ')
+                    c2 = ''
+                    if counts[2] == 'Crash!':
+                        c2 = 'Crash'
+                    c1 = ''
+                    if counts[1] == 'Crash!':
+                        c1 = 'Crash'
+                    html += fmt(package, datestr, c2, c1) + '\n'
+                    if c1 != 'Crash':
+                        break
+                if line.startswith('Program received signal '):
+                    crash_line = next(file, '')
+                    location_index = crash_line.rindex(' at ')
+                    if location_index > 0:
+                        crash_location = crash_line[location_index+3:].strip()
+                        if crash_location in crash_locations:
+                            crash_locations[crash_location] += 1
+                        else:
+                            crash_locations[crash_location] = 1
                     break
-                else:
-                    # Current package, parse on
-                    continue
-            if line.startswith(str(current_year) + '-') or line.startswith(str(current_year - 1) + '-'):
-                datestr = line
-            if not line.startswith('count:'):
-                continue
-            if line.find('Crash') < 0:
-                break
-            package = filename[filename.rfind('/')+1:]
-            counts = line.strip().split(' ')
-            c2 = ''
-            if counts[2] == 'Crash!':
-                c2 = 'Crash'
-            c1 = ''
-            if counts[1] == 'Crash!':
-                c1 = 'Crash'
-            html += fmt(package, datestr, c2, c1, None) + '\n'
-            break
+
+    html += '</pre>\n'
+    html += '<pre>\n'
+    html += '<b>' + fmt('HEAD Crash locations', 'Count', link=False) + '</b>\n'
+    for crash_location in sorted(crash_locations.items(), key=lambda (l, c): (-c, l)):
```

Comment:
This does not work any longer with Python 3:
```
0.09s$ python3 -m py_compile ./tools/donate-cpu-server.py
  File "./tools/donate-cpu-server.py", line 193
    for crash_location in sorted(crash_locations.items(), key=lambda (l, c): (-c, l)):
                                                                     ^
SyntaxError: invalid syntax
The command "python3 -m py_compile ./tools/donate-cpu-server.py" exited with 1.
```
Currently, the script does not work in Python 3, but at least it compiles :)
Regarding that issue, I also stumbled over it IIRC.
I guess it is this line: https://github.com/danmar/cppcheck/blob/d1fce6b7adee2552f53629720973c1c2959b3150/tools/donate-cpu-server.py#L644
Not sure if that works here too.

Reply:
Thanks, this seems to work:
```python
    for crash_location in sorted(crash_locations.items(), key=lambda x: (-x[1], x[0])):
```

Reply:
Actually, I've reworked it and this line has been rewritten.